### PR TITLE
[Advanced Compiler]Add Unfold backward

### PR DESCRIPTION
**Pull request by AdvancedCompiler**
### PR Category

Operator 
### Type of Change

New Feature
### Description

add Unfold backward
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

test:
4c4b1be49125e62f887c5305d4b05d9e
benchmark:
d8efe127900464daf6335c8560d3e417

### CODE REVIEW

File: src/flag_gems/__init__.py [hunk 1]
```diff
@@ -344,6 +344,7 @@ def torch_ge(v):
     ("true_divide.Tensor", true_divide),
     ("true_divide_.Scalar", true_divide_),
     ("true_divide_.Tensor", true_divide_),
+    ("unflod_backward", unfold_backward),
```
Comment (tengqm) [inline src/flag_gems/__init__.py]: check typo here.

File: src/flag_gems/ops/unfold_backward.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import torch
+import triton
+import triton.language as tl
+
+
+@triton.jit
+def _unfold_bwd_kernel(
```
Comment (tengqm) [inline src/flag_gems/ops/unfold_backward.py]: it is hard to guess that 'bwd' means 'backward'.
please consider rename this to `_unfold_backward_kernel` unless this naming is mandatory.

File: src/flag_gems/__init__.py (hunk 1 above)
Comment (Ea760) [reply src/flag_gems/__init__.py]: Fixed the issue already

File: src/flag_gems/ops/unfold_backward.py (hunk 2 above)
Comment (Ea760) [reply src/flag_gems/ops/unfold_backward.py]: Fixed the issue already

File: src/flag_gems/result.json [hunk 3]
```diff
@@ -0,0 +1,332 @@
+{
```
Comment (tengqm) [inline src/flag_gems/result.json]: Please remove this and the next result.json file from the PR.

Comment (Ea760) [reply src/flag_gems/result.json]: The file has been deleted in the latest commit.

Comment (Ea760) [reply src/flag_gems/result.json]: The CI/CD seems to have failed. Could a maintainer re-run the workflow? (Pushing again would add an unnecessary commit.）

Comment (tengqm) [reply src/flag_gems/result.json]: No. The `src/flag_gems/ops/result.json` file is still there...

Comment (Ea760) [reply src/flag_gems/result.json]: Sorry, I cleared the root directory json but missed the one in src/flag_gems.

File: benchmark/test_special_perf.py [hunk 4]
```diff
@@ -881,3 +881,42 @@ def input_kwargs(shape, dtype, device):
...
+
+    class UnfoldBackwardBenchmark(Benchmark):
+        def set_shapes(self, shape_file_path=None):
+            self.shapes = [
+                ((32, 64), 1, 16, 16),
+                ((16, 33), 0, 5, 2),
+                ((4, 8, 12), -1, 6, 4),
+                ((7, 13), 1, 13, 3),
+                ((6, 20), 1, 7, 4),
+                ((2, 3, 17), -1, 9, 1),
```
Comment (tengqm) [inline benchmark/test_special_perf.py:906]: Do you have some special considerations when selecting these shapes?

File: tests/test_special_ops.py [hunk 5]
```diff
@@ -1890,3 +1890,34 @@ def _verify_expert_level_sorting(
...
+@pytest.mark.label("unfold_backward")
+@pytest.mark.parametrize(
+    "input_sizes, dim, size, step",
+    [
+        ((32, 64), 1, 16, 16),
+        ((16, 33), 0, 5, 2),
+        ((4, 8, 12), -1, 6, 4),
+        ((7, 13), 1, 13, 3),
+        ((6, 20), 1, 7, 4),
+        ((2, 3, 17), -1, 9, 1),
```
Comment (tengqm) [inline tests/test_special_ops.py:1917]: Same question as above.

File: benchmark/test_special_perf.py (hunk 4 above)
Comment (Ea760) [reply benchmark/test_special_perf.py:906]: yes, these shape case can hit the important regimes: different tensor ranks (2D/3D), different dim choices (including -1), and different size/step patterns. We cover step == size (no overlap), several step < size cases with varying overlap (including step=1), and a boundary case where size == D so there’s only one window. i have add a step > size case. the correct test also pass.
image

File: tests/test_special_ops.py (hunk 5 above)
Comment (Ea760) [reply tests/test_special_ops.py:1917]: yes, these shape case can hit the important regimes: different tensor ranks (2D/3D), different dim choices (including -1), and different size/step patterns. We cover step == size (no overlap), several step < size cases with varying overlap (including step=1), and a boundary case where size == D so there’s only one window. i have add a step > size case. the correct test also pass.
image

Comment (tengqm) [reply tests/test_special_ops.py:1917]: Thanks for the clarification. Happy to know that these parameters were not chosen in an ad-hoc manner.

File: benchmark/test_special_perf.py (hunk 4 above)
Comment (tengqm) [reply benchmark/test_special_perf.py:906]: Thanks for the clarification. I was hoping that these parameters were designed with some implicit assumptions including some corner cases. Happy to know that was the case.

File: result.json [hunk 6]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (0x45f) [inline result.json]: remove this file

File: benchmark/test_special_perf.py [hunk 7]
```diff
@@ -881,3 +881,43 @@ def input_kwargs(shape, dtype, device):
     )
     bench.set_gems(flag_gems.per_token_group_quant_fp8)
     bench.run()
+
+
+@pytest.mark.unfold_backward
```
Comment (0x45f) [inline benchmark/test_special_perf.py]:
```suggestion
@pytest.mark.unfold
```

File: tests/test_special_ops.py [hunk 8]
```diff
@@ -1890,3 +1890,35 @@ def _verify_expert_level_sorting(
     gems_assert_close(
         num_tokens_post_pad, to_reference(num_tokens_post_pad_vllm), dtype=dtype
     )
+
+
+@pytest.mark.unfold_backward
```
Comment (0x45f) [inline tests/test_special_ops.py]:
```suggestion
@pytest.mark.unfold
```

File: tests/test_special_ops.py [hunk 9]
```diff
@@ -1890,3 +1890,35 @@ def _verify_expert_level_sorting(
...
+@pytest.mark.label("unfold_backward")
```
Comment (0x45f) [inline tests/test_special_ops.py]: remove this line

File: src/flag_gems/ops/unfold_backward.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    out_id = ((before_lin * D) + pos) * prod_after + after_lin
+
+    tl.atomic_add(grad_out_ptr + out_id, vals_f32, mask=mask)
+
+
+def unfold_backward(
+    grad_in: torch.Tensor, input_sizes, dim: int, size: int, step: int
+) -> torch.Tensor:
+    if step <= 0:
```
Comment (0x45f) [inline src/flag_gems/ops/unfold_backward.py:48]: add logger.debug

File: result.json (hunk 6 above)
Comment (Ea760) [reply result.json]: already remove

File: src/flag_gems/ops/unfold_backward.py (hunk 10 above)
Comment (Ea760) [reply src/flag_gems/ops/unfold_backward.py:48]: already add it
